internal/observability/events/context_test.go

```diff
@@ -0,0 +1,855 @@
+			}
+			op := tt.name
+			if tt.noOperation {
+				op = ""
+			}
+			require.Greater(len(tt.auditOpts), 0)
+			for _, opts := range tt.auditOpts {
+				opts := append(opts, event.WithNow(now))
+				err := event.WriteAudit(tt.ctx, event.Op(op), opts...)
+				if tt.wantErrMatch != nil {
+					require.Errorf(err, "wanted error %q", tt.wantErrMatch)
+					if tt.wantErrContains != "" {
+						assert.Contains(err.Error(), tt.wantErrContains)
+					}
+					return
+				}
+				require.NoError(err)
+			}
+			if !tt.noFlush {
+				require.NoError(event.WriteAudit(tt.ctx, event.Op(op), event.WithFlush(), event.WithNow(now)))
+			}
+			if tt.auditSinkFileName != "" {
+				defer func() { _ = os.WriteFile(tt.auditSinkFileName, nil, 0o666) }()
+
+				b, err := ioutil.ReadFile(tt.auditSinkFileName)
+				require.NoError(err)
+				gotAudit := &eventJson{}
+				err = json.Unmarshal(b, gotAudit)
+				require.NoErrorf(err, "json: %s", string(b))
+
+				actualJson, err := json.Marshal(gotAudit)
+				require.NoError(err)
+
+				wantEvent := eventJson{
+					CreatedAt: gotAudit.CreatedAt,
+					EventType: string(gotAudit.EventType),
+					Payload: map[string]interface{}{
+						"auth":            tt.wantAudit.Auth,
+						"id":              gotAudit.Payload["id"],
+						"timestamp":       now,
+						"request":         tt.wantAudit.Request,
+						"serialized_hmac": "",
+						"type":            apiRequest,
+						"version":         testAuditVersion,
+					},
+				}
+				if tt.wantAudit.Id != "" {
+					wantEvent.Payload["id"] = tt.wantAudit.Id
+					wantEvent.Payload["request_info"] = event.RequestInfo{
+						Id: tt.wantAudit.Id,
+					}
+				}
+				if tt.wantAudit.Response != nil {
+					wantEvent.Payload["response"] = tt.wantAudit.Response
+				}
+				wantJson, err := json.Marshal(wantEvent)
+				require.NoError(err)
+
+				assert.JSONEq(string(wantJson), string(actualJson))
+			}
+		})
+	}
+}
+
+func Test_WriteError(t *testing.T) {
+	// this test and its subtests cannot be run in parallel because of it's
+	// dependency on the sysEventer
+	now := time.Now()
+
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name: "test",
+	})
+
+	c := event.TestEventerConfig(t, "WriteAudit")
+
+	e, err := event.NewEventer(logger, c.EventerConfig, event.WithNow(now))
+	require.NoError(t, err)
+
+	info := &event.RequestInfo{Id: "867-5309"}
+
+	testCtx, err := event.NewEventerContext(context.Background(), e)
+	require.NoError(t, err)
+	testCtx, err = event.NewRequestInfoContext(testCtx, info)
+	require.NoError(t, err)
+
+	testCtxNoInfoId, err := event.NewEventerContext(context.Background(), e)
+	require.NoError(t, err)
+	noId := &event.RequestInfo{Id: "867-5309"}
+	testCtxNoInfoId, err = event.NewRequestInfoContext(testCtxNoInfoId, noId)
+	require.NoError(t, err)
+	noId.Id = ""
+
+	testError := errors.New(errors.InvalidParameter, "test", "you lookin at me?")
+
+	tests := []struct {
+		name            string
+		ctx             context.Context
+		e               error
```

Comment:
Could `e` be omitted here if we're just writing out `testError` in every test anyway?

Reply:
but we're not writing out `testError` we're writing `e`

Reply:
It just seemed like since every test case used `testError`, that it didn't necessarily need to be there. Not a huge deal though! :+1: